File: nowcoder/question/NC17.cpp
class Solution {
public:
    int getLongestPalindrome(string A, int n) {
        // write code here
        int dp[n][n];
        int maxLen = 1;
        int tempLen = 3;
        memset(dp, 0, sizeof(dp));

        if (n < 2) {
            return n;
        }

        for (int i = 0; i < n; ++i) {
            dp[i][i] = 1;
        }

        for (int i = 0; i < n - 1; ++i) {
            if (A[i] == A[i + 1]) {
                dp[i][i + 1] = 1;
                maxLen = 2;
            }
        }

        while (tempLen <= n) {
            for (int i = 0; i < n - tempLen + 1; ++i) {
                if (A[i] == A[i + tempLen - 1]) {
                    if (dp[i + 1][i + tempLen - 2]) {
                        dp[i][i + tempLen - 1] = 1;
                        maxLen = tempLen;
                    }
                }
                else {
                    dp[i][i + tempLen - 1] = 0;
                }
            }
            ++tempLen;
        }
        return maxLen;
    }
};

File: 滑动窗口/NC17.py
class Solution:
    def getLongestPalindrome(self, A, n):
        # write code here
        if n < 2:
            return n
        maxLen = 1
        for i in range(2, n + 1):
            index = 0
            while index + i <= n:
                if A[index:index + i] == A[index + i - 1 - n : index - 1 - n : -1]:
                    maxLen = i
                    break
                index += 1
        return maxLen